### ISSUE TITLE
Recursive types

### ISSUE THREAD

**Issue by nixpulvis** [egonSchiele/contracts.ruby#101]
It would be great to be able to write the following type `BinaryTree = Rec[:bt, Or[Num, Pair[:bt], Pair[:bt]]]`. Where `:bt` is the type of the whole binary tree type within the body of the definition. Using a symbol is just one option. But the general ability to define types like this is what I'm getting at.

**Comment by egonSchiele** [comment]
I'm not sure I follow, could you show a longer example?

**Comment by nixpulvis** [comment]
The binary tree example is a pretty simple example, but in essence the goal is to be able to refer to the type you are defining inside of itself. A binary tree for example is in my example above either a `Num` or a pair of `BinaryTree`. In a lazy language you might be able to write this like the following.

```ruby
BinaryTree = Or[Num, Pair[BinaryTree], Pair[BinaryTree, BinaryTree]]]
```

But in a strict language `BinaryTree` will not yet be defined so referencing it inside of the type definition here will not work.

Another more classic example of a recursive datatype is the list.

```ruby
NumList = Or[Empty, Cons[Num, NumList]]
```

**Comment by egonSchiele** [comment]
That's a limitation of ruby. Maybe you can achieve something similar with inheritance?

```ruby
class Tree
end

class Node < Tree
end

class Val < Tree
end

Contract Tree => nil
def traverse tree
```

**Comment by nixpulvis** [comment]
It is *not* a limitation of ruby, there are plenty of strict languages who implement recursive datatypes. Typed Racket or ML are great examples.

Typed Racket solves the problem by defining a `Rec` construct which effectively acts as a `letrec` in types. I think this same concept could be implemented in Ruby with symbols instead of constants.

**Comment by egonSchiele** [comment]
Ah, I see what you're saying now. Yeah, that's a good idea...would just need a new Rec contract and maybe a way to tie a symbol to a type.

**Comment by nixpulvis** [comment]
Exactly. The work done here if done in a smart way will also be able to open up doors for polymorphic types as well.

**Comment by alex-fedorov** [comment]
What about this syntax:

```ruby
BinaryTreeContract = Or[Num, Pair[Rec[:BinaryTreeContract], Rec[:BinaryTreeContract]]]
```

`Rec` will use something along the lines of `Object.const_get`

**Comment by alex-fedorov** [comment]
Another option would be generalise even more:

```ruby
Tree = Or[Ref[:value], Pair[Ref[:self], Ref[:self]]]
Tree.resolve_refs!(value: Num, self: Tree)
```

**Comment by alex-fedorov** [comment]
Unless, we do something like this:

```ruby
Tree = Or[Ref[:value], Pair[Ref[:self], Ref[:self]]].resolve_refs(value: Num, self: Self)
```

Where `Self = lambda { |obj| obj }` and if `resolve_refs` encounters a block/lambda it will call it with itself to get the value.

**Comment by nixpulvis** [comment]
I don't think people should be expected to "resolve" their own contracts. These matters should be dealt with internally when the contract system goes to check a guard. Calling it `Ref` is interesting, allowing for references to anything, not just the current contract being defined.

This will need to be done with some care however, consider the following contract.

```ruby
X = Ref[:X]
```

Regardless of the syntax chosen, this will cause issues. Recursive datatypes need to be shown to be productive, meaning that a recursive reference will always be smaller.

**Comment by nixpulvis** [comment]
Totally not calling this well thought out enough to be *done* but heres something that works.

```ruby
require 'contracts'
include Contracts

class Ref < CallableClass
  def initialize(contract)
    @contract = contract
  end

  def valid?(val)
    instance_eval(@contract).valid?(val)
  end

  def to_s
    @contract.to_s
  end
end

List = Or[nil, [String, Ref['List']]]

Contract List => Num
def size(list)
  if list == nil
    0
  else
    size(list[1]) + 1
  end
end

p size(nil)
p size(["hello", nil])
p size(["hello", ["world", nil]])
p size(["hello", [:bar, nil]])
```

**Comment by nixpulvis** [comment]
Another option is what the coffee script guys are doing.

```
o ::
  a: Num
  b: Self
  c: (Num) -> Self
  inner: { y: Bool, z: Self }
o = #...
```
> Self binds to the closest object contract. So in this example, Self in b and c points to o and Self in inner.z points to inner.

**Comment by vendethiel** [comment]
> Another option is what the coffeescript guys are doing.

Not sure what you're talking about, but there are no plans for a static typing/contract thing for coffee :).

**Comment by nixpulvis** [comment]
The coffeescript library for contracts.

**Comment by vendethiel** [comment]
Ah, alright, you're talking about [http://disnetdev.com/contracts.coffee/](this) (which is abandoned. The author, which also created sweet.js, moved on to [http://github.com/disnet/contracts.js](contracts.js)).

**Comment by waterlink** [comment]
To escape passing around symbols and strings, we can do something like that:

```ruby
Tree = Or[Value, Pair[Self, Self]]
```

`Self` is the special contract, that will keep track of top contract and be actual reference to it. Don't yet know how to implement it in most elegant way.

**Comment by waterlink** [comment]
Another solution I see - explicitly register your contracts into some kind of contract dictionary, and then `Ref[:Tree]` will just fetch it from this dictionary:

```ruby
Tree = Or[Value, Pair[Ref[:Tree], Ref[:Tree]]]
Contracts.register_ref(:Tree, Tree)
```

**Comment by nixpulvis** [comment]
I also was playing around with a `Self` however I feel it would be confusing because there is also `self`. I'd opt to calling the same concept `Rec` to draw parallels to typed racket.

**Comment by egonSchiele** [comment]
From chat by @waterlink:

You can generate class out of the `Rec[:X]`, which will have method X and it will be available in the context of block provided to `Rec`
So:

```
Rec[:X] { Or[Num, Pair[X, X]] }
```

will work without any *_missing and any rescue-ing. It will just call method :X on anonymous class returned by `Rec[:X] { .. }`

**Comment by nixpulvis** [comment]
Missing from the chat was the return value should be inside the Rec block.

**Comment by waterlink** [comment]
`Rec` should be just a builtin contract I think. Example:

```ruby
TreeContract = C::Rec[:X] { C::Or[Num, PairContract[X, X]] }
# ...
Contract TreeContract, TreeContract => TreeContract
def union(a, b)
  # ...
end
```

If you want to define it inline:

```ruby
Contract C::Num => C::Rec[:X, lambda { C::Or[Num, C::Pair[X, X]] }]
def generate_random_tree(seed)
  # ...
end
```
